### DISCUSSION TITLE
3-digit hex colors

### DISCUSSION THREAD

**Discussion by pmur002** [r-devel/r-project-sprint#73] [Ideas]
A [post on fosstodon](https://fosstodon.org/@coolbutuseless/110972696890598490) suggests adding support for `"#RGB"` colours in R (in addition to existing `"#RRGGBB"` colours.

This may not be an urgent need and might be a small change, but it would still require the complete modify, build, test, check, document process to be covered.  And that is assuming there are no unintended consequences ...

**Comment by malcolmbarrett** [comment]
I'm interested in this!

**Comment by EllaKaye** [comment]
Me too!

**Comment by pmur002** [comment]
Cool.  The first step is to look at the fosstodon post;  that shows the exact region of the C code that (probably) requires a change.  In order to attempt this change, you will need to be able to (at a minimum):
1. Get a copy of the R source
2. Build R from source
If you are happy doing this, then you could have a go at making the change and seeing if you can get R to build again.
The next steps would include:
3. Testing your change (designing a test example first)
4. Updating the documentation that refers to R colours.
5. Running 'make check-all' to check for side-effects.
If you would like to tackle as many of those steps as you can (I have attempted to make the list deliberately too long), then get back to me with your progress and questions if you get stuck, even if you get stuck at Step 1 :)
I hope that gives you something to try.
I will be able to answer questions in the morning (for me; overnight for you).
As I mentioned, we could also have a zoom session tomorrow morning (for you), as long as it is quite early (possibly overlapping with/instead of the scheduled report back and/or kick-off).

**Comment by coolbutuseless** [comment]
This was my suggestion.   It has come up a few times when I was interfacing R with html/css (which allows for 3-digit hex colours).

According to https://www.w3.org/TR/css-color-3/ and other sources, `#abc` is equivalent to `#aabbcc`  (not `#a0b0c0`)

I had a play around with adapting some colour lookup code here: https://github.com/coolbutuseless/colourlookup

Another idea in that repo is using hash lookup for R-color-name-to-rgba-values in `col2rgb()`.   This gave a big speed improvement, but definitely increased complexity over current linear search.

I'm also happy to help out over chat on mastodon if there are any questions.

**Comment by pmur002** [comment]
Thanks!  And apologies for not properly attributing you with the original idea!

**Comment by coolbutuseless** [comment]
@pmur002 No worries!   This looks like a great small task for the sprint!

**Comment by EllaKaye** [comment]
This has now become issue #74.
